lua/luasnip/loaders/from_snipmate.lua

```diff
@@ -0,0 +1,218 @@
+---@diagnostic disable: redefined-local
+local ls = require("luasnip")
+local caches = require("luasnip.loaders._caches")
+local util = require("luasnip.util.util")
+local Path = require("luasnip.util.path")
+local session = require("luasnip.session")
+
+local function parse_snipmate(buffer, filename)
+	local snippet = {}
+	local extends = {}
+
+	---@type string[]
+	local lines = vim.split(buffer, "\n")
+	local i = 1
+
+	local function _parse()
+		local line = lines[i]
+		local prefix = line:match("%s+(%S+)")
+		local description = line:match([["(.+)"%s*$]])
+		local body = {}
+
+		i = i + 1
+		---@type number
+		local indent
+
+		while i <= #lines do
+			line = lines[i]
+			if line:find("^%s+") then
+				if not indent then
+					indent = #line:match("^%s+")
+				end
+				line = line:sub(indent + 1)
+				line = line:gsub("${VISUAL}", "${TM_SELECTED_TEXT}")
+			elseif line ~= "" then
+				break
+			end
+			table.insert(body, line)
+			i = i + 1
+		end
+
+		body = table.concat(body, "\n")
+		local snip = ls.parser.parse_snippet({
+			trig = prefix,
+			dscr = description,
+			wordTrig = true,
+		}, body)
+		table.insert(snippet, snip)
+	end
+
+	while i <= #lines do
+		local line = lines[i]
+		if vim.startswith(line, "snippet") then
+			_parse()
+		elseif vim.startswith(line, "extends") then
+			local scopes = vim.split(vim.trim(line:sub(8)), "[,%s]+")
+			vim.list_extend(extends, scopes)
+			i = i + 1
+		elseif vim.startswith(line, "#") or line:find("^%s*$") then
+			-- comment and blank line
+			i = i + 1
+		else
+			error(("invalid line in %s: %s"):format(filename, i))
+		end
+	end
+
+	return snippet, extends
+end
+
+local function load_snippet_file(lang, path)
+	if not Path.exists(path) then
+		return
+	end
+
+	Path.async_read_file(
+		path,
+		vim.schedule_wrap(function(buffer)
+			local snippet, extends = parse_snipmate(buffer, path)
+			if not snippet then
+				return
+			end
+			table.insert(extends, lang)
+			for _, ft in ipairs(extends) do
+				local lang_snips = ls.snippets[ft] or {}
+				ls.snippets[ft] = vim.list_extend(lang_snips, snippet)
+				session.latest_load_ft = ft
+				vim.cmd("do User LuasnipSnippetsAdded")
+			end
+		end)
+	)
+end
+
+local function list2set(list)
+	vim.validate({ list = { list, "table", true } })
+	if not list then
+		return
+	end
+
+	local out = {}
+	for _, v in ipairs(list) do
+		out[v] = true
+	end
+	return out
+end
+
+local function filter(exclude, include)
+	vim.validate({
+		exclude = { exclude, "table" },
+		include = { include, "table", true },
+	})
+
+	exclude = list2set(exclude)
+	include = list2set(include)
+
+	return function(lang)
+		if exclude[lang] then
+			return false
+		end
+		if include == nil or include[lang] then
+			return true
+		end
+	end
+end
+
+local function get_paths(root)
+	local ret = {}
+	local files = Path.scandir(root, "file")
+	for _, file in ipairs(files) do
+		local name, ext = Path.basename(file, true)
+		if name == "_" then
+			name = "all"
+		end
```

Comment:
Not sure if I'm completely happy with handling it like this, the `_`-convention could be different for other snippet-collections.

Reply:
It would be easy to hard-code it to look at the parent directory name to determine if it is vim-snippets or not. But I wonder if that's enough.

Reply:
It's possible to `ls.filetype_extend("all", {"<whichever filetype is used by this snippet collection for all>"})` to search in `_` too for global snippets, that would make it pretty easy to modify for any source.
One downside is that `include` and `exclude` aren't aware of this `ft-redirection`, so we would have to change that.

Reply:
I tried to add it in line 152, but it seems to be different: neither include nor exclude is set, but _.snippets is not searched.